### ISSUE TITLE
Requesting cost of invalid extending instruction

### ISSUE THREAD

**Issue by dnsampaio** [llvm/llvm-project#64252]
At line
https://github.com/llvm/llvm-project/blob/fcb6a9c07cf7a2bc63d364e3b7f60aaadadd57cc/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L7745

the SLP vectorizer will request the cost of an invalid instruction performing an extension from a larger type to a smaller one whenever the slp decides that the reduction operation can be performed in lesser bits than the original type here:

https://github.com/llvm/llvm-project/blob/fcb6a9c07cf7a2bc63d364e3b7f60aaadadd57cc/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L7294

 Reproducer:
rep.ll
```
define i32 @c() {
entry:
  %rem.1 = srem i32 undef, 2
  %idxprom.1 = sext i32 %rem.1 to i64
  %arrayidx.1 = getelementptr inbounds i16, ptr @c, i64 %idxprom.1
  %rem.6 = srem i32 undef, 2
  %idxprom.6 = sext i32 %rem.6 to i64
  %arrayidx.6 = getelementptr inbounds i16, ptr @c, i64 %idxprom.6
  unreachable
}
```
opt --slp-vectorizer < rep.ll

**Comment by dnsampaio** [comment]
Our downstream target does a sanity check before discovering the cast instructions, the debug log output is:
```
*** IR Dump Before SLPVectorizerPass on c ***
define i32 @c() {
entry:
  %rem.1 = srem i32 undef, 2
  %idxprom.1 = sext i32 %rem.1 to i64
  %arrayidx.1 = getelementptr inbounds i16, ptr @c, i64 %idxprom.1
  %rem.6 = srem i32 undef, 2
  %idxprom.6 = sext i32 %rem.6 to i64
  %arrayidx.6 = getelementptr inbounds i16, ptr @c, i64 %idxprom.6
  unreachable
}

Features:
CPU:kv3-1
TuneCPU:kv3-1

SLP: Analyzing blocks in c.
SLP: Found GEPs for 1 underlying objects.
SLP: Analyzing a getelementptr list of length 2.
SLP: Trying to vectorize a list of length = 2.
SLP: Analyzing 2 operations 
SLP:  bundle:   %idxprom.1 = sext i32 %rem.1 to i64
SLP:  initialize schedule region to   %idxprom.1 = sext i32 %rem.1 to i64
SLP:  extend schedule region end to   %idxprom.6 = sext i32 %rem.6 to i64
SLP: try schedule bundle [  %idxprom.1 = sext i32 %rem.1 to i64;  %idxprom.6 = sext i32 %rem.6 to i64] in block entry
SLP:       update deps of [  %idxprom.1 = sext i32 %rem.1 to i64;  %idxprom.6 = sext i32 %rem.6 to i64]
SLP:       update deps of /   %idxprom.6 = sext i32 %rem.6 to i64
SLP:       update deps of   %arrayidx.1 = getelementptr inbounds i16, ptr @c, i64 %idxprom.1
SLP:     gets ready on update:   %arrayidx.1 = getelementptr inbounds i16, ptr @c, i64 %idxprom.1
SLP:    initially in ready list:   %arrayidx.1 = getelementptr inbounds i16, ptr @c, i64 %idxprom.1
SLP:   schedule   %arrayidx.1 = getelementptr inbounds i16, ptr @c, i64 %idxprom.1
SLP:    gets ready (def): [  %idxprom.1 = sext i32 %rem.1 to i64;  %idxprom.6 = sext i32 %rem.6 to i64]
SLP: We are able to schedule this bundle.
SLP: added a vector of casts.
SLP: We are able to schedule this bundle.
SLP: added a vector of un/bin op.
SLP: Gathering due to C,S,B,O, small shuffle. 
SLP: Gathering due to C,S,B,O, small shuffle. 
SLP: Checking user:  %arrayidx.1 = getelementptr inbounds i16, ptr @c, i64 %idxprom.1.
SLP: Need to extract:  %arrayidx.1 = getelementptr inbounds i16, ptr @c, i64 %idxprom.1 from lane 0 from   %idxprom.1 = sext i32 %rem.1 to i64.
SLP: Checking user:  %arrayidx.6 = getelementptr inbounds i16, ptr @c, i64 %idxprom.6.
SLP: Need to extract:  %arrayidx.6 = getelementptr inbounds i16, ptr @c, i64 %idxprom.6 from lane 1 from   %idxprom.6 = sext i32 %rem.6 to i64.
SLP: Checking user:  %idxprom.1 = sext i32 %rem.1 to i64.
SLP:    Internal user will be removed:  %idxprom.1 = sext i32 %rem.1 to i64.
SLP: Checking user:  %idxprom.6 = sext i32 %rem.6 to i64.
SLP:    Internal user will be removed:  %idxprom.6 = sext i32 %rem.6 to i64.
DemandedBits: Root:   unreachable
SLP: Calculating cost for tree of size 4.
Extended type is not larger than original.:<2 x i32> == (40) ==> <2 x i16>
```

**Comment by alexey-bataev** [comment]
Fixed in 0a68cd23049de778a4440168c855aff38f202940
